### PR TITLE
CASEFLOW-1499 - Do Not Allow User To Select Subsequent Hearings Docket Tasks

### PR DESCRIPTION
**Pull request by n1zyy**
### Description
Resolves [CASEFLOW-1499](https://vajira.max.gov/browse/CASEFLOW-1499).

### Acceptance Criteria
- [ ] Code compiles correctly

### Testing Plan
I added a new JSON-exported appeal that makes this easy to test. You have two options:

* If you are about to go grab lunch or something, pull in this PR and rebuild your database; it's added as a new seed.
* Based on the [Exporting and Importing Appeals](https://github.com/department-of-veterans-affairs/caseflow/wiki/Exporting-and-Importing-Appeals) page, import the appeal directly (see below)

To import directly:

```
require "helpers/sanitized_json_configuration.rb"
require "helpers/sanitized_json_importer.rb"
# Maximize verbosity to see what the importer is doing
sji = SanitizedJsonImporter.from_file("db/seeds/sanitized_json/b5eba21a-9baf-41a3-ac1c-08470c2b79c4.json", verbosity: 5)
sji.import
```

Then, visit /queue/appeals/b5eba21a-9baf-41a3-ac1c-08470c2b79c4 logged in as a Clerk of the Board user (e.g., COB_USER).

1. Press "Add Substitute," input a date, and select a claimant (choices here are not significant for this). Press Continue
2. On the next page, confirm that the "Distribution" task is checked and disabled (existing functionality), and...
3. Select "Evidence Submission Window" and confirm that "Schedule Hearing" is disabled (but not checked).
4. Uncheck "Evidence Submission Window" and confirm that Schedule Hearing is re-enabled
5. Check "Schedule Hearing" and confirm that "Evidence Submission Window" is re-enabled
6. In one of these states (remember which you select!), press "Continue"
7. Click through the review screen
8. Confirm that a new appeal with the substitute appellant is created, that the task you selected is carried over to the new appeal, and that the one that was disabled is not. (This is actually not part of this story.)

### User Facing Changes
 - [x] Screenshots of UI changes added to PR & Original Issue

Selecting "Evidence Submission Window" disables "Schedule hearing":

Screen Shot 2021-07-22 at 9 48 04 PM

And vice versa:

Screen Shot 2021-07-22 at 9 48 11 PM

### Storybook Story
* [ ] Test that this carries over into Storybook properly. (It should, but I haven't tested yet.)

### CODE REVIEW

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 1]
```diff
@@ -68,12 +68,63 @@ export const isDescendantOfDistributionTask = (taskId, taskList) => {
   return isDescendant(tasksById, distributionTask, task, { id: 'taskId' });
 };
 
+export const scheduleHearingTaskDisableTypes = [
+  'AssignHearingDispositionTask',
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:72]: This was referred to as "Select Hearing Disposition" in the Jira ticket. Want to confirm my mapping here is correct.

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 2]
```diff
@@ -68,12 +68,63 @@ export const isDescendantOfDistributionTask = (taskId, taskList) => {
...
+  'ChangeHearingDispositionTask',
+  'EvidenceSubmissionWindowTask',
+  'TranscriptionTask'
+];
+
+export const evidenceSubmissionWindowDisableTypes = [
+  'ScheduleHearingTask',
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:79]: Similarly, this was called "Schedule Veteran Task" in Jira, and want to make sure this is also the correct thing to use here.

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 3]
```diff
@@ -68,12 +68,63 @@ export const isDescendantOfDistributionTask = (taskId, taskList) => {
...
+
+export const transcriptionTaskDisableTypes = [
+  'ScheduleHearingTask',
+  'AssignHearingDispositionTask',
+  'ChangeHearingDispositionTask'
+];
+
+export const taskTypesSelected = ({ tasks, selectedTaskIds }) => {
+  return tasks.filter((task) => selectedTaskIds.includes(parseInt(task.taskId, 10))).map((task) => task.type);
+};
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:92]: This is duplicative of what's in [Paula's PR](https://github.com/department-of-veterans-affairs/caseflow/pull/16524/files) (it is, in fact, stolen from her PR), but I've moved code around above and below so there's a decent chance of a merge conflict. I am happy to help resolve it if I'm not the one that gets stuck with it.

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 4]
```diff
@@ -68,12 +68,63 @@ export const isDescendantOfDistributionTask = (taskId, taskList) => {
...
+
+export const shouldDisableBasedOnTaskType = (taskType, selectedTaskTypes) => {
+  // This is not efficient; it'll make multiple passes through each array.
+  // Realistically it's a small set, so get it working first and we can tweak later.
+  // We _currently_ don't need to handle multiple selections because the latter two types
+  // disable the same things. But this code can be improved.
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:98]: @jcq and I talked a fair bit about this.

This isn't code I'm particularly proud of. For every task in the list, we call this method, which looks through the array of task types potentially up to 3 times, and if it enters one of them, it will run through the array of related types.

*But*, it executes nearly instantaneously. It's unlikely that there will ever be more than perhaps 10 items to run through, so we're talking about a couple hundred iterations of a loop looking at stuff in memory, worst case. If we try to write this in the "efficient" way our brains want to, it gets really unreadable.

Separately: I initially erred in making this an `if ... else` construct. Multiple boxes can be checked, and we need to compute the total set to be excluded. But, by happenstance, this configuration works fine. EvidenceSubmissionWindowTask and TranscriptionTask disable the exact same fields, so the fact that we miss an iteration of "and then do the exact same thing again" doesn't matter. Selecting either of them will disable ScheduleHearingTask, so that can't be selected along with them. And if you instead select ScheduleHearingTask, it disables both of the other two ones.

So, tl;dr, this code is mildly embarrassing, but it also works fine. If someone has a thoughtful refactor I'm all ears, but for now I'd rather skew towards wasting some microseconds of CPU time versus inscrutable code.

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 5]
```diff
@@ -68,12 +68,63 @@ export const isDescendantOfDistributionTask = (taskId, taskList) => {
...
   return alwaysDisabled.includes(taskInfo.type);
 };
 
+export const disabledTasksBasedOnSelections = ({ tasks, selectedTaskIds }) => {
+  const selectedTaskTypes = taskTypesSelected({ tasks, selectedTaskIds });
+
+  return tasks.map((task) => {
+    return ({
+      ...task,
+      disabled: shouldDisable(task) || shouldDisableBasedOnTaskType(task.type, selectedTaskTypes)
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:123]: This initially read:

```ruby
disabled: task.disabled || shouldDisableBasedOnTaskType(task.type, selectedTaskTypes)
```

This turned out to be really problematic. Once a box was disabled, it would never be un-disabled. Calling `shouldDisable` gets us the expected behavior. (If some day there's a third way to disable boxes on the form, this will be a much bigger headache.)

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 6]
```diff
@@ -145,7 +196,7 @@ export const prepTaskDataForUi = (taskData, claimantPoa) => {
   return sortedTasks.map((taskInfo) => ({
     ...taskInfo,
     hidden: shouldHide(taskInfo, claimantPoa, taskData),
-    disabled: shouldDisable(taskInfo, taskData),
+    disabled: shouldDisable(taskInfo),
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:199]: This second param was never actually being used.

File: client/test/app/queue/substituteAppellant/tasks/utils.test.js [hunk 7]
```diff
@@ -76,6 +76,139 @@ describe('utility functions for task manipulation', () => {
...
+        expect(shouldDisableBasedOnTaskType(taskType, selectedTaskTypes)).toBe(true);
+      });
+
+      it.each(shouldNotDisables)('should not disable task type %s', (taskType) => {
+        expect(shouldDisableBasedOnTaskType(taskType, selectedTaskTypes)).toBe(false);
+      });
+    });
+  });
+
+  describe('disabledTasksBasedOnSelections', () => {
```
Comment (n1zyy) [inline client/test/app/queue/substituteAppellant/tasks/utils.test.js:154]: A handful of different thoughts here.

As my comment on the test for `disabledTasksBasedOnSelections` at the top of this section says, these two tests are _highly_ duplicative. Technically, one tests the boolean returned, and the other tests that the task was actually disabled. Is there any value in keeping both of them, or should we remove the first one, since any problems would also surface here?

Second, in both tests, the three `describe` conditions share very similar code. If this were Ruby where I'm much more comfortable refactoring, I'd have written a little method like `expectDisabled(array_of_task_types)` and `expectNotDisabled(array_of_task_types)`. While there is perhaps some value to very explicitly spelling out what we expect, it looks like a lot of copypasta to me. (Also: can confirm, the second and third describe blocks _are_ copypasta.)

File: client/test/app/queue/substituteAppellant/tasks/utils.test.js [hunk 8]
```diff
@@ -76,6 +76,139 @@ describe('utility functions for task manipulation', () => {
...
+    const tasks = [
+      { taskId: 1, type: 'EvidenceSubmissionWindowTask' },
+      { taskId: 2, type: 'ScheduleHearingTask' },
+      { taskId: 3, type: 'AssignHearingDispositionTask' },
+      { taskId: 4, type: 'ChangeHearingDispositionTask' },
+      { taskId: 5, type: 'TranscriptionTask' }
+    ];
+
+    describe('when ScheduleVeteranTask is selected', () => {
```
Comment (n1zyy) [inline client/test/app/queue/substituteAppellant/tasks/utils.test.js:163]:
```suggestion
    describe('when ScheduleHearingTask is selected', () => {
```

D'oh! Will fix in the AM.

File: spec/factories/appeal.rb [hunk 9]
```diff
@@ -245,6 +245,26 @@
...
+        appeal.veteran.update!(date_of_death: 1.month.ago)
+      end
+    end
+
+    trait :with_transcription_task do
+      after(:create) do |appeal, _evaluator|
+        root_task = RootTask.find_or_create_by!(appeal: appeal, assigned_to: Bva.singleton)
+        TranscriptionTask.create!(appeal: appeal, parent: root_task)
+      end
+    end
```
Comment (n1zyy) [inline spec/factories/appeal.rb:266]: This trait does not work and should be removed.

The other two are helpful and used in the testing instructions here. It's debatable just how much non-appeal stuff we want to cram into the Appeal factory. The way we talk about appeals, "Give me an appeal with an ESW task and a deceased veteran so we can test substitutions" makes a lot of sense, but I worry about this becoming a kitchen sink of weird things.

File: spec/feature/queue/substitute_appellant/substitute_appellant_spec.rb [hunk 10]
```diff
@@ -34,7 +34,7 @@
         it_should_behave_like "fill substitution form"
       end
 
-      context "with direct review docket" do
+      context "with hearing docket" do
```
Comment (n1zyy) [inline spec/feature/queue/substitute_appellant/substitute_appellant_spec.rb:37]: Unrelated fix: this came up while pairing.

**Comment by n1zyy** [comment]
This patch is currently adding a useless `disabled=""` attribute in an existing snapshot. Need to look into why in the morning.

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 11]
```diff
@@ -68,21 +68,68 @@ export const isDescendantOfDistributionTask = (taskId, taskList) => {
...
+
+export const transcriptionTaskDisableTypes = [
+  'ScheduleHearingTask',
+  'AssignHearingDispositionTask',
+  'ChangeHearingDispositionTask'
+];
+
+export const taskTypesSelected = ({ tasks, selectedTaskIds }) => {
+  return tasks.filter((task) => selectedTaskIds.includes(parseInt(task.taskId, 10))).map((task) => task.type);
+};
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:92]: This method is merely moved up from the old lines 82-85 because it's used below.

File: client/app/queue/substituteAppellant/tasks/utils.js [hunk 12]
```diff
@@ -68,21 +68,68 @@ export const isDescendantOfDistributionTask = (taskId, taskList) => {
...
+
 export const shouldHideBasedOnPoa = (taskInfo, claimantPoa) => {
   // eslint-disable-next-line camelcase
   return taskInfo.type === 'InformalHearingPresentationTask' && !claimantPoa?.ihp_allowed;
 };
 
-export const taskTypesSelected = ({ tasks, selectedTaskIds }) => {
-  return tasks.filter((task) => selectedTaskIds.includes(parseInt(task.taskId, 10))).map((task) => task.type);
-};
-
```
Comment (n1zyy) [inline client/app/queue/substituteAppellant/tasks/utils.js:85]: Just moved up to new lines 90-92.

File: client/test/app/queue/substituteAppellant/__snapshots__/SubstituteAppellantTasksForm.test.js.snap [hunk 13]
```diff
@@ -649,6 +650,7 @@ exports[`SubstituteAppellantTasksForm with existing form data renders default st
                     <input
                       aria-label="taskIds[2]"
                       checked=""
+                      disabled=""
```
Comment (n1zyy) [inline client/test/app/queue/substituteAppellant/__snapshots__/SubstituteAppellantTasksForm.test.js.snap:653]: I actually don't understand why this is happening, but I updated the snapshot since it _is_. 🤷

File: client/test/data/queue/substituteAppellant/tasks.js [hunk 14]
```diff
@@ -226,3 +226,27 @@ export const sampleTasksForEvidenceSubmissionDocket = () => {
...
+      label: 'Change hearing disposition',
+    },
+    {
+      type: 'TranscriptionTask',
+      label: 'Transcription task',
+    },
+  ];
+
+  return getAmaTasks(taskTypes);
+};
```
Comment (n1zyy) [inline client/test/data/queue/substituteAppellant/tasks.js:252]: I implemented this to just tack onto `sampleTasksForEvidenceSubmissionDocket()` because I'm returning an unrealistic set of tasks here to demo the form behavior in the storybook, but the existing method was used in multiple places.

I appear to have written this small bit of JS code correctly on my first attempt???

File: spec/factories/appeal.rb [hunk 15]
```diff
@@ -245,6 +245,19 @@
...
+        EvidenceSubmissionWindowTask.create!(appeal: appeal, parent: root_task)
+      end
+    end
+
+    trait :with_deceased_veteran do
+      after(:create) do |appeal, _evaluator|
+        appeal.veteran.update!(date_of_death: 1.month.ago)
+      end
+    end
+
```
Comment (n1zyy) [inline spec/factories/appeal.rb:260]: I'm having second thoughts about adding these.

They're handy because I wanted to use them for demo, and my instructions for manual review mention these. But they're not actually used in any code or tests. If it weren't _my_ code, I'd probably argue that adding code because we might want to use it some time later is probably a bad idea. Think I should remove these?

**Review by jcq — CHANGES_REQUESTED** [review]
Fully tested in Storybook, and LGTM

See note re the oddity of adding hearing-related tasks to a list of tasks from `evidence_submission` docket; maybe it might be good to not include the evidence submission ones (though you still need things like `DistributionTask`) just for clarity sake?

File: client/test/app/queue/substituteAppellant/tasks/utils.test.js [hunk 16]
```diff
@@ -76,6 +76,139 @@ describe('utility functions for task manipulation', () => {
...
+      it.each(shouldDisables)('should disable task type %s', (taskType) => {
+        expect(shouldDisableBasedOnTaskType(taskType, selectedTaskTypes)).toBe(true);
+      });
+
+      it.each(shouldNotDisables)('should not disable task type %s', (taskType) => {
+        expect(shouldDisableBasedOnTaskType(taskType, selectedTaskTypes)).toBe(false);
+      });
+    });
+
+    describe('when an EvidenceSubmissionWindow task is selected', () => {
```
Comment (barshirtcliff) [inline client/test/app/queue/substituteAppellant/tasks/utils.test.js:105]: while I appreciate thoroughness, this... is a lot.  the mechanics of the method you wrote work the same for all three types, so I think just one of these describe sections is adequate.  beyond that we're copying over the working code, verbosely.

File: client/test/app/queue/substituteAppellant/tasks/utils.test.js (hunk 7 above)
Comment (barshirtcliff) [reply client/test/app/queue/substituteAppellant/tasks/utils.test.js:154]: oh look, we had the same thought!  Yes remove bloat-test pls.

File: client/test/data/queue/substituteAppellant/tasks.js (hunk 14 above)
Comment (barshirtcliff) [reply client/test/data/queue/substituteAppellant/tasks.js:248]: it's test data, and the comment clarifies that it's unrealistic.  I think it's good.

File: spec/factories/appeal.rb [hunk 17]
```diff
@@ -245,6 +245,19 @@
...
+        root_task = RootTask.find_or_create_by!(appeal: appeal, assigned_to: Bva.singleton)
+        EvidenceSubmissionWindowTask.create!(appeal: appeal, parent: root_task)
+      end
+    end
+
+    trait :with_deceased_veteran do
+      after(:create) do |appeal, _evaluator|
+        appeal.veteran.update!(date_of_death: 1.month.ago)
+      end
+    end
```
[the remark this answers is not in the document]
Comment (barshirtcliff) [reply spec/factories/appeal.rb:259]: yus

File: spec/factories/appeal.rb (hunk 15 above)
Comment (barshirtcliff) [reply spec/factories/appeal.rb:260]: keep 'em, I'll use them.

**Review by barshirtcliff — APPROVED** [review]
lgtm!

File: client/test/data/queue/substituteAppellant/tasks.js (hunk 14 above)
Comment (n1zyy) [reply client/test/data/queue/substituteAppellant/tasks.js:248]: Oh, for historical context: I changed this method based on JC's comment; JC was right for the previous code, and now Bar is right in response to the updated code. :)

**Comment by n1zyy** [comment]
Just demoed to Ann-Marie and JC. The issue Bar spotted on Friday turned out to have been surprisingly hard to fix, but we've banged on it a bunch and it works. Yay!
